x-pack/plugins/apm/common/critical_path/get_critical_path.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type {
+  IWaterfall,
+  IWaterfallSpanOrTransaction,
+} from '../../public/components/app/transaction_details/waterfall_with_summary/waterfall_container/waterfall/waterfall_helpers/waterfall_helpers';
+import { CriticalPath, CriticalPathSegment } from './types';
+
+export function getCriticalPath(waterfall: IWaterfall): CriticalPath {
+  const segments: CriticalPathSegment[] = [];
+
+  function scan({
+    item,
+    start,
+    end,
+  }: {
+    item: IWaterfallSpanOrTransaction;
+    start: number;
+    end: number;
+  }): void {
+    segments.push({
+      self: false,
+      duration: end - start,
+      item,
+      offset: start,
+    });
+    const directChildren = waterfall.childrenByParentId[item.id];
+
+    if (directChildren && directChildren.length > 0) {
+      // We iterate over all the item's direct children. The one that
+      // ends last is the first item in the array.
+      const orderedChildren = directChildren.concat().sort((a, b) => {
+        const endTimeA = a.offset + a.skew + a.duration;
+        const endTimeB = b.offset + b.skew + b.duration;
+        return endTimeB - endTimeA;
+      });
+
+      // For each point in time, determine what child is on the critical path.
+      // We start scanning at the end. Once we've decided what the child on the
+      // critical path is, scan its children, from the start time of that span
+      // until the end. The next scan time is the start time of the child that was
+      // on the critical path.
+      let scanTime = end;
+
+      orderedChildren.forEach((child) => {
+        const normalizedChildStart = Math.max(child.offset + child.skew, start);
+        const childEnd = child.offset + child.skew + child.duration;
+
+        // if a span ends before the current scan time, use the current
+        // scan time as when the child ended. We don't want to scan further
+        // than the scan time. This prevents overlap in the critical path.
+        const normalizedChildEnd = Math.min(childEnd, scanTime);
+
+        const isOnCriticalPath = !(
+          // A span/tx is NOT on the critical path if:
+          // - The start time is equal to or greater than the current scan time.
+          // Otherwise, spans that started at the same time will all contribute to
+          // the critical path, but we only want one to contribute.
+          // - The span/tx ends before the start of the initial scan period.
+          // - The span ends _after_ the current scan time.
+          (
+            normalizedChildStart >= scanTime ||
+            normalizedChildEnd < start ||
+            childEnd > scanTime
```

Comment:
@AlexanderWert do you remember what the reasoning was here? I understand that a span cannot be on the critical path if it ends after its parent, because it implies that the parent doesn't care about it finishing, but this also eliminates spans that end after its sibling that was previously on the critical path. Which might be a valid reason, but wanted to clarify that with you.

Reply:
Let's take this one as an example:

<img width="1368" alt="image" src="https://user-images.githubusercontent.com/866830/197458625-86f94eae-5030-41c9-b877-740ce9404aa5.png">

The algorithm starts from the end with `GET /api/order` and we know that `GET /api/order` is on the CP. The spans `Parse the document` and `Longtask` are partially parallel to `GET /api/order` because they are not parent of those, started before but DID NOT END before `GET /api/order`. This implies that the start of `GET /api/oder` is blocked by its parent rather one of the other two mentioned spans. As a consequence, the CP before `GET /api/order` lies on the parent `orders` and we can eliminate `Parce the document` and `Longtask` from the critical path.

Reply:
Thank you, that makes sense!